Add selected text execution like REPL

velios opened the pull request.

Commit:
Add selected text execution like REPL

sisoe24 commented on src/executer.ts:
```diff
@@ -0,0 +1,59 @@
+        return null;
+    }
+
+    const selectedText = editor?.document.getText(editor.selection);
```
Here, we can safely assume that `editor` will always have a value since we check for `null` in advance. In js/ts the `var?.var?` operator is called the Optional Chaining Operator and it is used to safely access properties or call methods on an object that might be `null` or `undefined`.

sisoe24 commented on src/executer.ts:
```diff
@@ -0,0 +1,59 @@
+    const shellCommand = `hs -c "${luaScript}"`;
+    let output = runSync(shellCommand);
+
+    if (output !== null) {
```
This is just a thought, but using an early return here might improve readability.  `if output === null; return null` so the chunk logic would not be indented and it would make it clear what is the outcome of the `output` variable. What do you think?

velios replied:
It definitely improves readability. Thank you for giving hints, I'm still not very good at what rules of good form when writing in typescript

velios replied:
It definitely improves readability. Thank you for giving hints, I'm still not very good at what rules of good form when writing in typescript

sisoe24 commented:
Don't worry about the static checks. They are out of date, and I should have taken them off. The code looks very good, and I left some minor comments that are more stylistic than anything.

It would be wise to test the `escapeString` function at this point. Tests are a little bit more complex to write but if you are up to the task I give some starting points:

```ts
// src/suite/executor.test.ts

import * as assert from "assert";

// remember to export the function
import { escapeString } from "../../executer";

suite("Test Escape String", () => {
    test("Test string with dobule quotes", async () => {
        assert.ok(escapEstring("hs.alert.show("name")") === "hs.alert.show(\"name\")");
    });
   
    test("Test string with single quotes", async () => {...})
});


```

Edit: if you want to test the extension, you can switch the debugging mode from "Run Extension" to "Extension Tests". To do this, just go to the Run and Debug Panel (shift cmd D). And if you want to focus only on your tests, you can isolate the testing suite with the `only` method:

```ts
suite.only(...);

// same applies for skip
suite("foo", () => {
    test.skip(....)};
)};

Commit:
Lint with Prettier

Commit:
Remove unnecessary nil check

Commit:
Add ut for escapeString fn

Commit:
Remove unnecessary comments

velios commented:
Thank you for guidance. I really appreciate you way of doing things. It's a little tricky to do tests, because of extra escaping added by javascript. But anyway now i sure it works pretty well.

P.S I can't pass one of test on my local machine. I don't understand how to fix it and why this happens.
```javascript
  1) Spoons Creation
       Create Spoon directory:
     Error: ENOENT: no such file or directory, rmdir '~/Code/hammerspoon-vscode/.tests/.hammerspoon/Spoons/test.spoon'
  	at Object.rmdirSync (node:fs:1219:10)
  	at Context.<anonymous> (~/Code/hammerspoon-vscode/out/test/suite/spoon.test.js:32:12)
  	at process.processImmediate (node:internal/timers:466:21)
  	at process.callbackTrampoline (node:internal/async_hooks:130:17)
```

Maybe i need to create new issue. I don't know.

Commit:
Add executeCurrentFile and executeCurrentLine

Commit:
Renaming execute to evaluate and executer to repl

Commit:
Avoid console steal focus

velios commented:
Add all planed modes. From selection, from current line and from whole file. Change some terminology, similar as in [Calva](https://github.com/BetterThanTomorrow/calva/tree/published) project. I believe it should name repl, even if we speak about send code to hammerspoon console and then process should named as evaluate, not execute. And I move some common code to new utilities namespace. May be it was wrong decision for this project, but i also take idea from Calva repo. It seems convinient for me.   

I added all the planned modes. Execute the selected text, the current line, and the entire file. Changed some terminology, following the example of the [Calva](https://github.com/BetterThanTomorrow/calva/) project. I think it should be called repl, even though we're talking about submitting code to the Hammerspoon console, and the process itself should be called evaluation, not execution. I also moved some common code into the new utilities namespace. It's ultimately up to you to decide if this is correct, but I took the idea and code from the Calva repository. I found it convenient. I am happy that my proposal was implemented so quickly. I have already started using the new functionality in my local build and am very happy that now it can be done. My Hammerspoon scripting has never been so enjoyable. Thank you for taking the time to teach me.

Commit:
Fix wrong regex

Commit:
Fix output window scroll when new output appended

sisoe24 reviewed (changes requested):
I'm curious about why you don't want to focus on the output window. Just in case it's helpful, there's a setting in the extension called `hammerspoon.console.focusOutputWindow`. It's a boolean setting that decides whether or not the output console is focused on when the Hammerspoon config is reloaded. Maybe we can use this setting in this scenario as well.

sisoe24 commented:
Great job! I left you some comments we can discuss further.

Commit:
Add changes after review

velios commented:
I made the suggested changes. Of course, there is still a lot that needs to be improved. I already tried using it locally today to work with Hammerspoon. It's a wonderful experience, I'm very satisfied. Perhaps I will make more suggestions as I work with the plugin in practice.

Commit:
One more commit

velios commented:
Changes requested is pending. Do I have to click something else to move the process further inside the github?

velios commented:
Changes requested is pending. Do I have to click something else to move the process further inside the github?

sisoe24 commented on package.json:
It would be a good idea to consider putting the menu entry in a sub-menu for better organization.

```json
"submenus": [
      {
        "id": "hammerspoon.repl",
        "title": "Hammerspoon",
        "when": "editorTextFocus && resourceLangId == lua",
        "group": "hammerspoon",
        "label": "Hammerspoon"
      }
 ],
"menus": {
      "editor/title": [
        {
          "when": "resourceLangId == lua",
          "command": "hammerspoon.reloadConfiguration",
          "group": "hammerspoon"
        }
      ],
      "editor/context": [
        {
          "submenu": "hammerspoon.repl",
          "group": "hammerspoon"
        }
      ],
      "hammerspoon.repl": [
        {
          "when": "resourceLangId == lua",
          "command": "hammerspoon.reloadConfiguration",
          "group": "hammerspoon"
        },
        {
          "command": "hammerspoon.evaluateCurrentFileText",
          "group": "hammerspoon",
          "when": "editorTextFocus && resourceLangId == lua"
        },
        {
          "command": "hammerspoon.evaluateSelectedText",
          "group": "hammerspoon",
          "when": "editorTextFocus && resourceLangId == lua"
        },
        {
          "command": "hammerspoon.evaluateCurrentLineText",
          "group": "hammerspoon",
          "when": "editorTextFocus && resourceLangId == lua"
        }
      ]
    }

```

sisoe24 commented on src/utilities.ts:
Would it be better to combine the two functions into one and relocate it to where it's needed? Usually, I use the 'utils' module for functions that don't have a specific place yet. But since this function is only used by the REPL, it makes sense to put it there.

Also, I'm not quite sure if we should throw an error if the text editor isn't focused. It might not be very useful to the user since they are the ones calling the code. Plus, we'd need to double-check for errors when using this function. Personally, I think it would be better to just exit quietly in this situation. What do you think about that?

velios replied:
```typescript
const editor = vscode.window.activeTextEditor;
    if (editor) {...}
```
I see the following pattern not only in my namespaces, but also in Spoons.ts and lua_parser.ts, so a generalization seemed appropriate to me. I used a high order function, but perhaps decorators should have been used here. Or just repeat this piece of code from time to time. What do you think about that? In any case, I implemented a solution that seemed good to me and uploaded the code. If you have time, please take a look.

Commit:
Refactor repl ns

Commit:
Add submenu and fmt with prettifier

velios commented:
I was hoping you would use squash&merge option to my commit. Now develop branch a little messy.

sisoe24 commented:
You are right. I didn't think about that. I don't know if reverting now is a good idea.
